### PR TITLE
Fix z-schema module definition and add SchemaErrorDetail

### PR DESCRIPTION
**Pull request by octogonz**
Changes made in this PR:
- Removed "export as namespace ZSchema" because there is no global variable
- Replaced the "export = ZSchema.Validator" with a module declaration so that we can access the other types such as ZSchema.Options
- Fixed a problem where ZSchema.SchemaError was including properties that are actually from a completely different data structure, which I named SchemaErrorDetail
- Added support for SchemaError.details and SchemaErrorDetail.inner, which is needed for rich error reporting
- Added JSDoc explanations for these structures
- The unit test file now actually uses these types, rather than relying on implicit types

Please fill in this template.
- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/zaggino/z-schema
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@andy-ms Can you please take a look at the test that didn't run at all.

**Comment by pgonzal** [comment]
@sheetalkamat I'm not sure why travis is failing.  "tsc" is able to compile z-schema-tests.ts on my local box.

**Comment by job13er** [comment]
I don't actually use typescript anymore, any way to switch who is automatically flagged to review changes here?

**Comment by andy-ms** [comment]
We don't accept pull requests to `types-2.0` any more (#13233). Please rebase onto `master`, then hit "Edit" in the upper-right and change the base branch to `master`.

**Comment by pgonzal** [comment]
@job13er  I'd volunteer to be listed as the contact.  Our projects use z-schema extensively.

**Comment by pgonzal** [comment]
@sheetalkamat / @andy-ms is there anything else I need to do before merging this?

Also, per Adam's question, how is the review contact determined?  Is it parsed from the code comments at the top of index.d.ts?

**Comment by andy-ms** [comment]
The reviewer is determined from the `// Definitions by: Adam Meadows <https://github.com/job13er>` comment. You can add yourself to the end to become a reviewer.
@job13er You can make an issue on https://github.com/DefinitelyTyped/dt-review-tool if you don't want to be notified any more.

**Comment by pgonzal** [comment]
@job13er do you want me to remove your name from the `// Definitions by:` line when I add my name there?

**Comment by job13er** [comment]
Works for me, thanks

File: z-schema/index.d.ts
Comment (mhegazy) [inline z-schema/index.d.ts]: do not use `module` use `namespace`; they both mean the something, except that `namespace` is the preferred keyword now.

File: z-schema/index.d.ts
Comment (mhegazy) [inline z-schema/index.d.ts]: do not declare an ambient external module. just leave the `export=` at the bottom of the file.

**Review by mhegazy — CHANGES_REQUESTED** [review]

File: z-schema/index.d.ts
Comment (pgonzal) [inline z-schema/index.d.ts]: @mhegazy  Could you give an example?  I originally tried this style:

```typescript
declare namespace ZSchema {
    export interface SchemaError extends Error {
        . . .
    }
    . . .
    export class Validator {
       . . .
    }
}
export = ZSchema.Validator;
```

...but I could not get it to work.  For example:

```typescript
import Validator = require('z-schema');
var validator: Validator = new Validator(options);
var error: ZSchema.SchemaError = validator.getLastError(); // error: Cannot find namespace ZSchema
```

Since the "import" statement brings in the "=" export, it seems like I would need some other expression to import the namespace, something like this:

```typescript
import Validator = require('z-schema');
import { ZSchema } from 'z-schema';
```

...but it doesn't work.  The ambient definition is the only syntax I could find that allows me to access the ZSchema namespace definitions.

I'd greatly appreciate any help/suggestions for a better solution.

Comment (andy-ms) [inline z-schema/index.d.ts]: If you're writing in a module declaration, namespaces aren't global. But you could declare a namespace to be merged with the class:
```ts
declare namespace Validator {
    interface SchemaError extends Error { ... }
}
declare class Validator { ... }
export = Validator;
```
then:
```ts
import Validator = require("z-schema");
var error: Validator.SchemaError = new Validator(options).getLastError();
```

Comment (pgonzal) [inline z-schema/index.d.ts]: Awesome, thanks, that's a much better solution!  I have pushed an update.  @mhegazy let me know if there are any remaining issues blocking this PR.

**Comment by pgonzal** [comment]
@mhegazy @andy-ms Hey guys, somehow GitHub seems to have deleted the recent conversation history.  I believe I addressed all your feedback; if so, please approve this PR.  Thanks!

For reference, here's the missing commentary:

**Andy 1/17/2017 5:15pm:**
> If you're writing in a module declaration, namespaces aren't global. But you could declare a namespace > to be merged with the class:
> declare namespace Validator {
>     interface SchemaError extends Error { ... }
> }
> declare class Validator { ... }
> export = Validator;
> then:
> import Validator = require("z-schema");
> var error: Validator.SchemaError = new Validator(options).getLastError();

**Pete 1/17/2017**
> Awesome, thanks, that's a much better solution! I have pushed an update.
> @mhegazy let me know if there are any remaining issues blocking this PR.

**Comment by andy-ms** [comment]
It's just hidden. Click the "Show outdated" button on the collapsed comments to view them.
Anyway, looks good from my end.

**Comment by pgonzal** [comment]
Sounds good.  @mhegazy can you approve this PR?

**Pull request merged**
